torch/_dynamo/variables/torch.py

```diff
@@ -738,6 +738,86 @@ def handle_ntuple(value):
             return handle_ntuple(args[0])
 
 
+def is_fn_safe_to_run(tx, f, sub_args):
+    # Snapshot state
+    graph_checkpoint, checkpoint = tx.output.graph, tx.copy_graphstate()
+    # Will raise if not sound
+    output, graph, lifted_freevars = speculate_subgraph(
+        tx, f, sub_args, graph_checkpoint, checkpoint, always_restore=True
+    )
+    # If we got here, we are probably sound, but we do not support freevars yet, so lets call cases where we found them
+    # unsafe.
+    if lifted_freevars:
+        return False
+    return True
+
+
+# See NOTE [HigherOrderOperator tracing design] for details of the design
+# See NOTE [speculate_subgraph vs old_speculate_subgraph] for other info
+def speculate_subgraph(
+    tx, f, sub_args, graph_checkpoint, checkpoint, *, always_restore=False
+):
+    from . import AutogradFunctionContextVariable, ConstantVariable, TensorVariable
+
+    try:
+        with tx.output.new_subtracer() as tracer:
+            args = []
+            # One argument to graph per sub_args
+            for a in sub_args:
+                assert not isinstance(
+                    a, torch.Tensor
+                ), "Tensors should already be tracked?"
+                if isinstance(a, ConstantVariable):
+                    proxy = tracer.create_graph_input("const")
```

Comment:
Are ConstVariable always Tensors or can they be other things?

Reply:
They should never be tensors, they are things like strings, ints, etc.